pkg/sysfs/parsers.go

```diff
@@ -115,6 +134,145 @@ func parseNumeric(path, value string, ptr interface{}) error {
+	if err != nil {
+		return 0.0, sysfsError(path, "can't parse numeric value '%s': %v", val, err)
+	}
+	return num * float64(unit), nil
+}
+
+// parseNumericMap parses a numeric string into integer of the right size.
+func parseNumericMap(path, key, value string, vmap interface{}) error {
+	var numstr string
+	var num, unit int64
+	var unum uint64
+	var f float64
+	var err error
+
+	if numstr, unit, err = splitNumericAndUnit(path, value); err != nil {
+		return err
+	}
+
+	switch vm := vmap.(type) {
+	case map[string]int8:
+		typeName, min, max := "int8", int64(math.MinInt8), int64(math.MaxInt8)
+		if num, err = parseIntWithUnit(path, numstr, unit); err != nil {
+			return err
+		}
+		if num < min || num > max {
+			return sysfsError(path, "numeric value %s overflows %s", value, typeName)
+		}
+		vm[key] = int8(num)
+	case map[string]uint8:
+		typeName, max := "uint8", uint64(math.MaxUint8)
+		if unum, err = parseUintWithUnit(path, numstr, unit); err != nil {
+			return err
+		}
+		if unum > max {
+			return sysfsError(path, "numeric value %s overflows %s", value, typeName)
+		}
+		vm[key] = uint8(unum)
+	case map[string]int16:
+		typeName, min, max := "int16", int64(math.MinInt16), int64(math.MaxInt16)
+		if num, err = parseIntWithUnit(path, numstr, unit); err != nil {
+			return err
+		}
+		if num < min || num > max {
+			return sysfsError(path, "numeric value %s overflows %s", value, typeName)
+		}
+		vm[key] = int16(num)
+	case map[string]uint16:
+		typeName, max := "uint16", uint64(math.MaxUint16)
+		if unum, err = parseUintWithUnit(path, numstr, unit); err != nil {
+			return err
+		}
+		if unum > max {
+			return sysfsError(path, "numeric value %s overflows %s", value, typeName)
+		}
+		vm[key] = uint16(unum)
+	case map[string]int32:
+		typeName, min, max := "int32", int64(math.MinInt32), int64(math.MaxInt32)
+		if num, err = parseIntWithUnit(path, numstr, unit); err != nil {
+			return err
+		}
+		if num < min || num > max {
+			return sysfsError(path, "numeric value %s overflows %s", value, typeName)
+		}
+		vm[key] = int32(num)
+	case map[string]uint32:
+		typeName, max := "uint32", uint64(math.MaxUint32)
+		if unum, err = parseUintWithUnit(path, numstr, unit); err != nil {
+			return err
+		}
+		if unum > max {
+			return sysfsError(path, "numeric value %s overflows %s", value, typeName)
+		}
+		vm[key] = uint32(unum)
+	case map[string]int64:
+		typeName, min, max := "int64", int64(math.MinInt64), int64(math.MaxInt64)
+		if num, err = parseIntWithUnit(path, numstr, unit); err != nil {
+			return err
+		}
+		if num < min || num > max {
+			return sysfsError(path, "numeric value %s overflows %s", value, typeName)
+		}
+		vm[key] = num
+	case map[string]uint64:
+		typeName, max := "uint64", uint64(math.MaxUint64)
+		if unum, err = parseUintWithUnit(path, numstr, unit); err != nil {
+			return err
+		}
+		if unum > max {
+			return sysfsError(path, "numeric value %s overflows %s", value, typeName)
+		}
+		vm[key] = unum
```

Comment:
There's a lot of duplicate code. Just a thought: could this be simplified by doing the limits checking in `parseUintWithUnit()` / `parseIntWithUnit()`? Either by giving the number of bits or min/max limits as arguments. WDYT?

Reply:
Yes, that's probably a good idea. I'll change it accordingly.